library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/api/RegistryDict.java

```diff
@@ -0,0 +1,338 @@
+		 * @param source the source to check
+		 * @return if the dictionary value should be loaded or not
+		 */
+		public boolean shouldLoad(ResourceType source) {
+			return this.source == source;
+		}
+	}
+
+	/**
+	 * Builder for dictionaries.
+	 *
+	 * @param <R> type of the entries in the registry
+	 * @param <V> attached value type
+	 */
+	final class Builder<R, V> {
+		private final Registry<R> registry;
+		private final Identifier id;
+		private final Class<V> valueClass;
+		private final Codec<V> codec;
+
+		private Side side;
+		private @Nullable V defaultValue;
+		private @Nullable ComputeFunction<R, V> computeFunction;
+
+		private Builder(Registry<R> registry, Identifier id, Class<V> valueClass, Codec<V> codec) {
+			this.registry = registry;
+			this.id = id;
+			this.valueClass = valueClass;
+			this.codec = codec;
+			side = Side.BOTH;
+
+			try {
+				if (get(registry, id, valueClass).isPresent()) {
+					throw new IllegalStateException("Dictionary with ID '%s' is already registered for registry %s!"
+							.formatted(id, registry.getKey().getValue()));
+				}
+			} catch (WrongValueClassException e) {
+				throw new IllegalStateException(("Dictionary with ID '%s' is already registered for registry %s, " +
+						"but with a different value class (expected %s, got %s)!")
+						.formatted(id, registry.getKey().getValue(), e.getExpectedClass(), e.getActualClass()));
+			}
+		}
+
+		/**
+		 * Sets what side this dictionary should exist on.
+		 *
+		 * @param side dictionary side
+		 * @return this builder
+		 */
+		public Builder<R, V> side(Side side) {
+			this.side = side;
+			return this;
+		}
+
+		/**
+		 * Sets the default value of this dictionary.
+		 *
+		 * <p>Setting this will <b>remove</b> the currently set
+		 * {@linkplain #computeFunction(ComputeFunction) compute function}!
+		 *
+		 * @param defaultValue default value
+		 * @return this builder
+		 */
+		public Builder<R, V> defaultValue(@Nullable V defaultValue) {
+			this.defaultValue = defaultValue;
+			this.computeFunction = null;
+			return this;
+		}
+
+		/**
+		 * Sets the <em>compute function</em> of this dictionary, which will be used to compute a value for a
+		 * specific entry, should it be missing.
+		 *
+		 * <p>Note that this will be computed on both sides and the computation result will <em>not</em> be synchronized.
+		 *
+		 * <p>Setting this will <b>remove</b> the currently set
+		 * {@linkplain #defaultValue(Object) default value}!
+		 *
+		 * @param computeFunction function to compute otherwise-missing value
+		 * @return this builder
+		 */
+		public Builder<R, V> computeFunction(@Nullable ComputeFunction<R, V> computeFunction) {
+			this.computeFunction = computeFunction;
+			this.defaultValue = null;
+			return this;
+		}
+
+		/**
+		 * Builds a new dictionary.
+		 *
+		 * @return new dictionary
+		 */
+		public RegistryDict<R, V> build() {
+			var dict = new RegistryDictImpl<>(registry, id, valueClass, codec, side, defaultValue, computeFunction);
```

Comment:
A single implementation having both the default value and compute function capabilities seems strange to me. Not a huge deal, but I might break this into two different implementations (possible with a shared abstract parent class).